ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -758,36 +767,45 @@ static void HandleKernelRequest(void* messageMemory, uint32_t messageSize)
                 requestHeader,
                 relativePath,
                 absolutePath,
+                nullptr, /* relativeFromPath */
                 requestHeader->messageType == MessageType_KtoU_NotifyDirectoryPreDelete,  // isDirectory
                 KUMessageTypeToNotificationType(static_cast<MessageType>(requestHeader->messageType)));
             break;
         }
         
         case MessageType_KtoU_NotifyFileHardLinkCreated:
         {
-#if DEBUG
-            // TODO(Mac): Move the following line out of the DEBUG block once we actually need the information. Currently just causes warning-as-error in release build.
-            const char* fromPath = request.paths[MessagePath_From];
-            const char* relativeFromPath = fromPath != nullptr ? GetRelativePath(fromPath, s_virtualizationRootFullPath.c_str()) : nullptr;
-
-            cout << "PrjFSLib.HandleKernelRequest: hard-linked " << (fromPath ?: "[out-of-root path]") << " -> " << (absolutePath ?: "[out-of-root path]") << " (absolute), ";
-            if (relativeFromPath != nullptr)
-            {
-                cout << "from this root (relative path " << relativeFromPath << ") ";
-            }
-            if (relativePath != nullptr)
+            const char* absoluteFromPath = request.paths[MessagePath_From];
+            const char* relativeFromPath = GetRelativePath(absoluteFromPath, s_virtualizationRootFullPath.c_str());
+            if (relativeFromPath == nullptr)
             {
-                cout << "into this root (relative path " << relativePath << ")";
+                goto CleanupAndReturn;
             }
+            
+#ifdef DEBUG
+            cout << "PrjFSLib.HandleKernelRequest: " << "hard-linked ";
+            cout << "from this root (relative path " << relativeFromPath << ") ";
+            cout << "into this root (relative path " << relativePath << ")";
             cout << endl;
 #endif
             
-            if (relativePath != nullptr)
+            if (strcmp(relativePath, "") == 0)
+            {
+                result = HandleFileNotification(
+                    requestHeader,
+                    relativePath,
+                    absolutePath,
+                    relativeFromPath,
+                    false, // isDirectory; TODO: may yet be a directory
```

Comment:
Not your change but I think we can remove this `TODO`.  It was added when `MessageType_KtoU_NotifyFileHardLinkCreated` could be called for directories, and it no longer will be.

Reply:
Since you can do a hardlink for a directory I think it's "OK" to leave.

It may not be a current requirement, but maybe in the future someday?